src/analysis/mesh/qgsmeshtriangulation.cpp

```diff
@@ -0,0 +1,316 @@
+    feedback->setProgress( 0 );
+
+  QgsCoordinateTransform transform( vectorLayer->crs(), mCrs, transformContext );
+
+  QgsFeatureIterator fIt = vectorLayer->getFeatures();
+  QgsFeature feat;
+  bool isZValueGeom = valueAttribute < 0;
+  int i = 0;
+  int featureCount = vectorLayer->featureCount();
+  while ( fIt.nextFeature( feat ) )
+  {
+    if ( feedback )
+    {
+      if ( feedback->isCanceled() )
+        break;
+
+      feedback->setProgress( 100 * i / featureCount );
+      i++;
+    }
+
+    QgsGeometry geom = feat.geometry();
+    geom.transform( transform, QgsCoordinateTransform::ForwardTransform, true );
+    QgsAbstractGeometry::vertex_iterator vit = geom.vertices_begin();
+
+    double value = 0;
+    if ( !isZValueGeom )
+      value = feat.attribute( valueAttribute ).toDouble();
+
+    while ( vit != geom.vertices_end() )
+    {
+      if ( feedback && feedback->isCanceled() )
+        break;
+      if ( isZValueGeom )
+        mTriangulation->addPoint( *vit );
+      else
+      {
+        ;
+        mTriangulation->addPoint( QgsPoint( QgsWkbTypes::PointZ, ( *vit ).x(), ( *vit ).y(), value ) );
+      }
+      ++vit;
+    }
+  }
+
+  return true;
+}
+
+bool QgsMeshTriangulation::addBreakLines( QgsVectorLayer *vectorLayer, int valueAttribute, const QgsCoordinateTransformContext &transformContext, QgsFeedback *feedback )
+{
+  if ( !vectorLayer )
+    return false;
+
+  QgsCoordinateTransform transform( vectorLayer->crs(), mCrs, transformContext );
+
+  QgsWkbTypes::GeometryType geomType = vectorLayer->geometryType();
+
+  switch ( geomType )
+  {
+    case QgsWkbTypes::PointGeometry:
+      return addVertices( vectorLayer, valueAttribute, transformContext, feedback );
+      break;
+    case QgsWkbTypes::LineGeometry:
+    case QgsWkbTypes::PolygonGeometry:
+    {
+      QgsFeatureIterator fIt = vectorLayer->getFeatures();
+      QgsFeature feat;
+      while ( fIt.nextFeature( feat ) )
+      {
+        addBreakLinesFromFeature( feat, valueAttribute, transform, feedback );
+      }
+      return true;
+    }
+    break;
+    default:
+      return false;
+  }
+
+  return false;
+}
+
+QgsMesh QgsMeshTriangulation::triangulatedMesh() const
+{
+  return mTriangulation->triangulationToMesh();
+}
+
+void QgsMeshTriangulation::setCrs( const QgsCoordinateReferenceSystem &crs )
+{
+  mCrs = crs;
+}
+
+void QgsMeshTriangulation::addBreakLinesFromFeature( const QgsFeature &feature, int valueAttribute, const QgsCoordinateTransform &transform, QgsFeedback *feedback )
+{
+  double valueOnVertex = 0;
+  if ( valueAttribute >= 0 )
+    valueOnVertex = feature.attribute( valueAttribute ).toDouble();
+
+  if ( feedback )
+    feedback->setProgress( 0 );
+
+  //from QgsTinInterpolator::insertData()
+  std::vector<const QgsCurve *> curves;
+  QgsGeometry geom = feature.geometry();
+  geom.transform( transform, QgsCoordinateTransform::ForwardTransform, true );
+  if ( QgsWkbTypes::geometryType( geom.wkbType() ) == QgsWkbTypes::PolygonGeometry )
```

Comment:
switch statement would be preferable here, so that it's explicit which logic applies for each different geometry type